Redemption timeout

tomaszslabon opened the pull request:
This PR adds functionality of redemption timeout notification.

One function was added: `notifyRedemptionTimeout`.
The function checks whether the redemption request exists and has timed out.
Further steps depend on the state of the wallet:
* if the wallet is in `Live` state:
  - the wallet's pending redemptions value is decreased by the amount requested in redemption request (minus treasury fee)
  - the tokens taken by the bank during the redemption request are returned to the redeemer
  - the redemption request is moved from pending redemptions to timeout redemptions
  - the wallet's operators are slashed (`TODO`)
  - the wallet is marked as either `MovingFunds` (if there is mainUTXO set for the wallet) or `Closed` (if there is no main UTXO set for the wallet)
  - if the wallet was marked as active - it is no longer marked as such
  
* if the wallet is in `MovingFunds` state:
  - the wallet's pending redemptions value is decreased by the amount requested in redemption request (minus treasury fee)
  - the tokens taken by the bank during the redemption request are returned to the redeemer
  - the redemption request is moved from pending redemptions to timeout redemptions
  - the wallet's operators are slashed (`TODO`)
  
* if the wallet is in `Terminated` state:
   - the wallet's pending redemptions value is decreased by the amount requested in redemption request (minus treasury fee)
  - the tokens taken by the bank during the redemption request are returned to the redeemer
  - the redemption request is moved from pending redemptions to timeout redemptions
  
* if the wallet is in `Unknown` or `Closed` state:
  - revert

The cost of calling `notifyRedemptionTimeout` is between `108k` and  `128k` (`116k` on average).

michalinacienciala commented on solidity/contracts/bridge/Bridge.sol:
```diff
@@ -1918,28 +1923,81 @@ contract Bridge is Ownable, EcdsaWalletOwner {
+            // Propagate timeout consequences to the wallet
+            wallets.notifyRedemptionTimedOut(walletPubKeyHash);
+        }
+        // TODO: should the notifier be reimbursed?
```
No, according to https://github.com/keep-network/tbtc-v2/blob/main/docs/rfc/rfc-6.adoc#notify-redemption-timeout. Notifier should be rewarded some % of slashed tokens.

tomaszslabon replied:
Done in 0e9b80110b1479820dfc3cbefcdf5054bf36b3df.

lukasz-zimnoch commented:
I think we can also remove the Slither-related TODO above `timedOutRedemptions` mapping

tomaszslabon commented on solidity/contracts/bridge/Bridge.sol:
```diff
@@ -389,6 +389,11 @@ contract Bridge is Ownable, EcdsaWalletOwner {
         bytes32 redemptionTxHash
     );
 
+    event RedemptionTimeout(
```
Done in 0e9b80110b1479820dfc3cbefcdf5054bf36b3df.

tomaszslabon commented on solidity/contracts/bridge/Bridge.sol:
```diff
@@ -1918,28 +1923,81 @@ contract Bridge is Ownable, EcdsaWalletOwner {
+        require(request.requestedAt > 0, "Redemption request does not exist");
+        require(
+            /* solhint-disable-next-line not-rely-on-time */
+            request.requestedAt + redemptionTimeout < block.timestamp,
```
Done in 0e9b80110b1479820dfc3cbefcdf5054bf36b3df.

tomaszslabon commented on solidity/contracts/bridge/Bridge.sol:
```diff
@@ -1918,28 +1923,81 @@ contract Bridge is Ownable, EcdsaWalletOwner {
+        // `requestRedemption` blocks requests targeting non-live wallets.
+        // Because `notifyRedemptionTimeout` changes the state of any live-wallet
+        // after the first call, there is no possibility that the given
+        // redemption key could be reported as timed-out multiple times.
```
Done in 427accdebc1b7e4003870b8016393dedc1e1886b.

tomaszslabon commented on solidity/contracts/bridge/Wallets.sol:
```diff
@@ -374,12 +375,6 @@ library Wallets {
-            walletState == WalletState.Live ||
-                walletState == WalletState.MovingFunds,
-            "ECDSA wallet must be in Live or MovingFunds state"
-        );
```
Brought back the check in 1594927c897618a89975336ff4dc730fb812db3b.

tomaszslabon commented on solidity/contracts/bridge/Bridge.sol:
```diff
@@ -1918,28 +1923,81 @@ contract Bridge is Ownable, EcdsaWalletOwner {
+            request.treasuryFee;
+
+        // Return the requested amount of tokens to the redeemer
+        bank.transferBalance(request.redeemer, request.requestedAmount);
```
We read the `request` as a pointer (and not the copy) to `pendingRedemption` element:
```
RedemptionRequest storage request = pendingRedemptions[redemptionKey];
```

And then we delete, so the `request` is zeroed:
```
delete pendingRedemptions[redemptionKey];
```

So we cannot move 
```
bank.transferBalance(request.redeemer, request.requestedAmount);
```
after deletion.

I would either leave it as is or work on a copy of the `request`.

tomaszslabon replied:
Moved the `bank.transferBalance(...)` at the end of the function (but before the even emitting).
I had to copy the request (use `memory` instead of `storage`).
 It decreased the gas consumption of `notifyRedemptionTimeout` a little bit (~1k).

tomaszslabon commented on solidity/contracts/bridge/Bridge.sol:
```diff
@@ -1918,28 +1923,81 @@ contract Bridge is Ownable, EcdsaWalletOwner {
+        // At the same time, if the given redemption key was already marked as
+        // fraudulent due to an amount-related fraud, it will not be possible
+        // to report a timeout on it since it will not be present in
+        // `pendingRedemptions` mapping.
```
I removed it.

tomaszslabon commented on solidity/contracts/bridge/Bridge.sol:
```diff
@@ -1918,28 +1923,81 @@ contract Bridge is Ownable, EcdsaWalletOwner {
+        );
+
+        // Update the wallet's pending redemptions value
+        Wallets.Wallet storage wallet = wallets.registeredWallets[
```
Done in d961863c5cc357ae6b1e77302d1dfed4c2ba9eed.

tomaszslabon commented:
> I think we can also remove the Slither-related TODO above `timedOutRedemptions` mapping

Done in 643921ac56f4d671f0975ef076f76a71ed171ab3.

lukasz-zimnoch commented on solidity/test/bridge/Bridge.test.ts:
```diff
@@ -4849,6 +4849,833 @@ describe("Bridge", () => {
+              await restoreSnapshot()
+            })
+
+            it("should not delete the active wallet public key hash", async () => {
```
That makes sense to not duplicate all the assertions in this scenario but let's drop a comment about it.

tomaszslabon replied:
Done in a769909bce54058695ce018258de2ada00bb6175.

lukasz-zimnoch commented on solidity/test/bridge/Bridge.test.ts:
```diff
@@ -4849,6 +4849,833 @@ describe("Bridge", () => {
+            expect(request.requestedAmount).to.be.equal(0)
+            expect(request.treasuryFee).to.be.equal(0)
+            expect(request.txMaxFee).to.be.equal(0)
+            expect(request.requestedAt).to.be.equal(0)
```
It's enough to check just if `requestedAt` is zero.

tomaszslabon replied:
Done in https://github.com/keep-network/tbtc-v2/commit/a769909bce54058695ce018258de2ada00bb6175.

lukasz-zimnoch commented on solidity/test/bridge/Bridge.test.ts:
```diff
@@ -4849,6 +4849,833 @@ describe("Bridge", () => {
+                expect(request.requestedAmount).to.be.equal(0)
+                expect(request.treasuryFee).to.be.equal(0)
+                expect(request.txMaxFee).to.be.equal(0)
+                expect(request.requestedAt).to.be.equal(0)
```
It's enough to check just if `requestedAt` is zero.

tomaszslabon replied:
Done in https://github.com/keep-network/tbtc-v2/commit/a769909bce54058695ce018258de2ada00bb6175.

lukasz-zimnoch commented on solidity/test/bridge/Bridge.test.ts:
```diff
@@ -4849,6 +4849,833 @@ describe("Bridge", () => {
+                expect(request.requestedAmount).to.be.equal(0)
+                expect(request.treasuryFee).to.be.equal(0)
+                expect(request.txMaxFee).to.be.equal(0)
+                expect(request.requestedAt).to.be.equal(0)
```
It's enough to check just if `requestedAt` is zero.

tomaszslabon replied:
Done in https://github.com/keep-network/tbtc-v2/commit/a769909bce54058695ce018258de2ada00bb6175.

lukasz-zimnoch commented on solidity/test/bridge/Bridge.test.ts:
```diff
@@ -4849,6 +4849,833 @@ describe("Bridge", () => {
+
+              it("should not emit WalletClosed events", async () => {
+                // Since the wallet has UTXO set it should be not be closed but
+                // marked as zMovingFunds` and we should not see any `WalletClosed`
```
Typo `zMovingFunds`

tomaszslabon replied:
Done in https://github.com/keep-network/tbtc-v2/commit/a769909bce54058695ce018258de2ada00bb6175.

lukasz-zimnoch reviewed (commented):
A couple of comments regarding tests

lukasz-zimnoch commented:
We have modified the state transition diagram (cc @michalinacienciala @pdyraga) and there are updates related to the redemption timeouts. However, I think we can just drop TODOs in relevant places and merge that PR as is because the changes will have a broader scope and it would be good to handle them in separate PRs.

tomaszslabon commented on solidity/test/bridge/Bridge.test.ts:
```diff
@@ -4849,6 +4849,833 @@ describe("Bridge", () => {
+          })
+        })
+
+        context("when the wallet is neither in Unknown or Closed state", () => {
```
Yes, done in ad9b362618f288407819c45df2676a51a38f13c9.

tomaszslabon commented on solidity/test/bridge/Bridge.test.ts:
```diff
@@ -4849,6 +4849,833 @@ describe("Bridge", () => {
+          })
+        })
+
+        context("when the wallet is in MovingFunds state", () => {
```
Yes, done in https://github.com/keep-network/tbtc-v2/commit/ad9b362618f288407819c45df2676a51a38f13c9.

tomaszslabon commented:
> We have modified the state transition diagram (cc @michalinacienciala @pdyraga) and there are updates related to the redemption timeouts. However, I think we can just drop TODOs in relevant places and merge that PR as is because the changes will have a broader scope and it would be good to handle them in separate PRs.

Added TODOs in 43f8606e05a4325f09a26fbe8b1d8bdadcc56be2.

lukasz-zimnoch reviewed (approved):
LGTM! Leaving final decision for @pdyraga

pdyraga commented on solidity/contracts/bridge/Bridge.sol:
```diff
@@ -1918,28 +1920,83 @@ contract Bridge is Ownable, EcdsaWalletOwner {
+    ///         `keccak256(walletPubKeyHash | redeemerOutputScript)`.
+    ///         The results of calling this function: the pending redemptions
+    ///         value for the wallet will be decreased by the requested amount
+    ///         (minus treasury fee), the tokens taken from the redeemer on
+    ///         redemption request will be returned to the redeemer, the request
```
No tokens are returned, the bank balance is returned.

```suggestion
    ///         (minus treasury fee), the balance taken from the redeemer on
    ///         redemption request will be returned to the redeemer, the request
```

pdyraga commented on solidity/hardhat.config.ts:
```diff
@@ -101,7 +101,7 @@ const config: HardhatUserConfig = {
     runOnCompile: true,
     strict: true,
-    except: ["BridgeStub$"], // TODO: Remove it once the problem of BridgeStub's size is solved
+    except: ["BridgeStub$", "Bridge$"], // TODO: Remove it once the problem of contracts' size is solved
```
😬

pdyraga commented:
I was not reviewing tests, trusting @lukasz-zimnoch opinion on them. Solidity looks good to me. There is one change in the documentation to `notifyRedemptionTimeout` we need to do - see my comment - but I do not consider it blocking, let's address it in the next PR.
